### ISSUE TITLE
add activeProperty to Screen

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/joist#418]
Suggested by @jonathanolson at 4/27/16 dev meeting:

Add `isActiveProperty` to Screen, would be useful for clients.

**Comment by pixelzoom** [comment]
Reminder to instrument `Screen.isActiveProperty` for PhET-iO.

**Comment by pixelzoom** [comment]
Skype discussion about desired behavior of isActiveProperty:

> [4/28/17, 2:29:04 PM] Chris Malley: Opinions on Screen.isActiveProperty...  Should its value change before or after the Screen's visibility changes?
[4/28/17, 2:29:12 PM] Chris Malley: I was thinking after.
[4/28/17, 2:29:42 PM] Sam Reid: It should become active right before becoming visible
[4/28/17, 2:29:58 PM] Sam Reid: and after becoming invisible it should become inactive
[4/28/17, 2:30:01 PM] Jonathan Olson: No strong preference
[4/28/17, 2:30:13 PM] Chris Malley: what SR described is going to be messy.
[4/28/17, 2:30:27 PM] Sam Reid: True, but it will avoid visual hiccups

So the desired behavior is:

• value transitions from false to true before a Screen becomes visible
• value transitions from true to false after a Screen becomes invisible.

**Comment by pixelzoom** [comment]
I thought this was going to be easy.  But after spending over an hour on it, I give up. Sim.js is a bloated fragile mess.  Trying to get the desired behavior of isActiveProperty at startup is unnecessarily complicated because Screens are all visible when initialized.  I had to touch too many things, and things started breaking.

**Comment by pixelzoom** [comment]
I took another stab at it, pushed above.

In Screen, there are 3 notable changes:

(1) Added `isActiveProperty`
(2) Added `assert` to verify that `isActiveProperty` doesn't change while the Screen is visible
(3) In `initializeView`, made the view invisible immediate after `createView` is called, and before it's added to the scenegraph

In Sim, all changes are in the function passed to `Property.multilink( [ this.showHomeScreenProperty, this.screenIndexProperty ],...)`. 

I tested with `graphing-lines_en.html?ea`, `graphing-lines_en.html?ea&screens=1` and `graphing-lines_en.html?ea&showHomeScreen=false`.

@samreid or @jonathanolson please review. High priority, since this affects all sims.

**Comment by pixelzoom** [comment]
I added PhET-iO instrumentation to Screen.isActiveProperty:

```js
    this.isActiveProperty = new Property( false, {
      tandem: options.tandem.createTandem( 'isActiveProperty' ),
      phetioValueType: TBoolean
    } );
```

I attempted to test in BLL instance-proxies:

![screenshot_19](https://cloud.githubusercontent.com/assets/3046552/25549193/bdcc3350-2c2d-11e7-89c7-691a26531e50.png)

When I toggle the "selected" check box, I see in the console:

```
Assertion failed: isActive should not change while the Screen view is visible
assert.js:20 Assertion failed: mismatched messageIndex, possible start/end mismatch
assert.js:21 Uncaught Error: Assertion failed: mismatched messageIndex, possible start/end mismatch
    at window.assertions.assertFunction (assert.js:21)
    at Object.end (phetioEvents.js?bust=1493417487598:154)
    at SimIFrameAPI.js?bust=1493417487598:265
```

The problem here is that we're attempting to change `isActiveProperty` while the Screen is visible, which is a no no.   Imo, we should not instrument `isActiveProperty`, it's not something that clients should have access to.  @samreid do you agree?

**Comment by samreid** [comment]
Yes, it seems like an implementation detail that should either

(a) not be instrumented

or

(b) have clear documentation that it is for observing only and not for setting (like a derived property.) . We have a few other properties like this in other sims, see `phetioInstanceDocumentation`.

Since I can envision clients that may want to detect when screens are active, I'm leaning toward (b).  Though a PhET-iO client would also be able to observe the screenIndexProperty, the isActiveProperty would be better.

**Comment by pixelzoom** [comment]
Ah... I just removed the instrumentation.  I'll restore it and figure out how to add `phetioInstanceDocumentation`.

**Comment by pixelzoom** [comment]
Instrumentation restored, `phetioInstanceDocumentation` added:

```js
    this.isActiveProperty = new Property( false, {
      tandem: options.tandem.createTandem( 'isActiveProperty' ),
      phetioValueType: TBoolean,
      phetioInstanceDocumentation: 'this Property is read-only, do not attempt to set its value'
    } );
```

instance-proxies still provides a "selected" checkbox that's editable.

**Comment by samreid** [comment]
I added this test harness:

```js
    this.isActiveProperty.link( function( isActive ) {
      console.log( options.name + ' ' + isActive );
    } );
```

And saw the printouts are correct for a 1-screen sim and a multi-screen sim. In the latter case, I saw:

```js
‪Concentration‬ false
‪Beer's Law‬ false
null false
‪
> tap screen 1
Concentration‬ true

> tap screen 2
‪Concentration‬ false
‪Beer's Law‬ true

> homescreen
‪Beer's Law‬ false
```

Changes in Sim.js and Screen.js look good.  I also tested BLL with ?screens=1 and it worked as expected.  

But I'm wondering how a screen model or view would get access to this new property?  We should probably pass it through as an option to `createModel` and `createView`, right?

**Comment by pixelzoom** [comment]
It's `@public`.  After calling `Screen.call`, any Screen subtype has access to it.

**Comment by pixelzoom** [comment]
Ah, I see... How to pass it to the view and model, since their instantiation is deferred.

**Comment by pixelzoom** [comment]
It should probably be an argument to `createModel` and `createView`, not an option.  Then the questions are (1) how to propagate it to the instances created by those functions, and (2) whether to propagate it all the way to ScreenView.

**Comment by pixelzoom** [comment]
I'm starting to question whether we should bail on this entire feature.

**Comment by samreid** [comment]
The reason I suggested it should be an option (passed like `{isScreenActiveProperty:...}`) to createModel and createView is so that, if we later add other parameters, the order won't seem odd, and clients that don't need the 1st arg can use the other args easily (without exposing unused 1st arg).

**Comment by pixelzoom** [comment]
What about propagation to ScreenView?  It's not needed currently, but might be the right thing to do now? If we do propagate, then we'll need to touch every ScreenView subtype (133 occurrences of `var ScreenView`), or default `options.isActiveProperty` to `null` in ScreenView.

**Comment by samreid** [comment]
I don't think it's necessary to pass to ScreenView.  I'm picturing something like ConcentrationScreen.js looking like this:

```js
    Screen.call( this,
      function() {
        return new ConcentrationModel( tandem.createTandem( 'model' ) );
      },
      function( model, options ) {
        return new ConcentrationView( model, modelViewTransform, options.isActiveProperty, tandem.createTandem( 'view' ) );
      },
      options
    );
```

That way each view can use it as desired or not at all--there would be no need to pass it through to ScreenView.

**Comment by pixelzoom** [comment]
I've renamed `isActiveProperty` to `activeProperty`.  This seems more consistent with other `{Property.<boolean>}` fields used throughout common code.  E.g. enabledProperty, onProperty, Sim.activeProperty.

**Comment by pixelzoom** [comment]
Example usage in example-sim.ExampleScreen (not committed):

```js
    Screen.call( this,
      function( options ) {
        return new ExampleModel( options.activeProperty );
      },
      function( model, options ) {
        return new ExampleScreenView( model, options.activeProperty );
      }, {
        backgroundColorProperty: new Property( 'rgb(50,50,50)' )
      }
    );
```

... then modify ExampleModel and ExampleScreenView constructors accordingly.

**Comment by pixelzoom** [comment]
The part that feels a little incorrect about this converting an option to a required function arg to the `createModel` and `createView` functions.  E.g. the above `createModel` function:  

```js
      function( options ) {
        return new ExampleModel( options.activeProperty );
      },
```

`options` is not really optional here, since it's required to correctly call `createModel` and `createView`.  

Back to @samreid for comment.

**Comment by pixelzoom** [comment]
So I guess my preference would be to pass `activeProperty` to `createModel` and `createView` as a required parameter, like:

```js
    Screen.call( this,
      function( activeProperty ) {
        return new ExampleModel( activeProperty );
      },
      function( model, activeProperty ) {
        return new ExampleScreenView( model, activeProperty );
      }, {
        backgroundColorProperty: new Property( 'rgb(50,50,50)' )
      }
    );
```

**Comment by samreid** [comment]
My concern about making it a required parameter, is that I suspect within the next year we'll be adding another argument.  Then many applications won't be using the first argument and usages will look like this:

```js
    Screen.call( this,
      function( activeProperty, soundEnabledProperty ) {
        return new ExampleModel( soundEnabledProperty );
      },
      function( model, activeProperty ) {
        return new ExampleScreenView( model, soundEnabledProperty );
      }, {
        backgroundColorProperty: new Property( 'rgb(50,50,50)' )
      }
    );
```

It makes sense to me that `options` is sometimes required to make things behave properly.  For instance to make a red button, passing in `fill:red` is not optional.  Another possibility would be to create a type like `ScreenModel` which has attributes like `screenModel.activeProperty`.

On the other hand, if we add an optional argument (of type Property.<boolean>) it could be fine for now and we could revisit in the future if/when we want to add more arguments to this call.

TLDR: I don't feel strongly about which pattern is used here at the moment but wanted to raise potential pros and cons for discussion.

**Comment by pixelzoom** [comment]
I would be more comfortable with:

```js
Screen.call( this,
  function( options ) { return new ExampleModel( options ); },
  function( model, options ) { return new ExampleScreenView( model, options ); }, 
  options );
```

... where it's up to `ExampleModel` and `ExampleScreenView` to unpack the options and decide whether to use `activeProperty`?

**Comment by samreid** [comment]
How about instead of cherry picking options, we pass the entire `Screen` and it is up to the model and view creation functions to choose what they need?

```js
Screen.call( this,
  function( screen ) { return new ExampleModel( screen.activeProperty ); },
  function( model, screen ) { return new ExampleScreenView( model, screen.soundEnabledProperty ); }, 
  options );
```

This will provide the cleanest decoupling from the model/view and screen, while still being very general.

**Comment by pixelzoom** [comment]
I kind of like this. But if we're passing the Screen, then the `model` argument to `createView` is unnecessary.

**Comment by pixelzoom** [comment]
I'm also concerned that this will encourage people to write code that _increases_ coupling to the Screen, e.g.:

```
Screen.call( this,
  function( screen ) { 
    return new ExampleModel( screen ); 
  },
  function( model, screen ) { return new ExampleScreenView( model, screen ); }, 
  options );
```

The `Screen` should definitely _not_ be passed to the model and view constructors.

**Comment by samreid** [comment]
To clarify, you are saying we would use `screen.model` instead of just `model`.  That seems reasonable.

> I'm also concerned that this will encourage people to write code that _increases_ coupling to the Screen

True, we would need to add a code review checklist item to make sure that doesn't happen.

Also, something that smells fishy is that this code is often *already* in a Screen declaration.  For instance, in ConcentrationScreen, we have

```js
    Screen.call( this,
      function() { return new ConcentrationModel( tandem.createTandem( 'model' ) ); },
      function( model ) { return new ConcentrationView( model, modelViewTransform, tandem.createTandem( 'view' ) ); },
      options );
```

We already have the context of the screen by using `self` and hence it seems odd to pass it as an argument.  Would it work to do this?

```js
var self = this;
Screen.call( this,
  function( screen ) { 
    return new ExampleModel( self.activeProperty ); 
  },
  function( model ) { return new ExampleScreenView( model, self.soundEnabledProperty ); }, 
  options );
```

??

**Comment by pixelzoom** [comment]
Ah yes, we could just rely on a closure to pass args to model and view constructors.

**Comment by pixelzoom** [comment]
I verified that this works just fine:

```js
    var self = this;
    Screen.call( this,
      function() {
        return new ExampleModel( self.activeProperty );
      },
      function( model ) {
        return new ExampleScreenView( model, self.activeProperty );
      },
      options );
```

@samreid OK with this solution?

**Comment by samreid** [comment]
Yes, looks good.
